app-header/app-header.html

```diff
@@ -45,184 +57,283 @@
+     * A cached offsetHeight of the current element.
+     * @type {number}
+     */
+    _height: 0,
+
+    /**
+     * The distance in pixels the header will be translated to when scrolling.
+     * @type {number}
+     */
+    _deltaHeight: 0,
+
+    /**
+     * The top most app-toolbar.
+     * @type {Element}
+     */
+    _topbar: null,
+
+    /**
+     *
+     * @type number
+     */
+    _oldY: 0,
+
+    /**
+     *
+     * @type number
+     */
+    _oldProgress: 0,
+
+    /**
+     *
+     * @type number
+     */
+    get _headerDy() {
+      return this.fixed ? this._deltaHeight : this._height + 5;
+    },
+
+    created: function() {
+      this._oldY = 0;
+    },
+
+    /**
+     * Recalculates the size of the toolbar.
+     *
+     * @method setup
+     */
+    setup: function() {
+      // noop if the header isn't visible
+      if (this.offsetWidth === 0 && this.offsetHeight === 0) {
+        return;
+      }
+      var firstSetup = this._height === 0;
+      var scrollTop = this._clampedScrollTop;
+      var currentNoTransition = this.noTransitions;
+      this._height = this.offsetHeight;
+      this._topbar = Polymer.dom(this).querySelector('app-toolbar');
 
-      _scrollHandler: function() {
-        this._updateHeader(this._clampedScrollTop);
-      },
+      // prepare for measurement
+      this.noTransitions = true;
 
-      _updateHeader: function(scrollTop) {
-        var y = 0;
-
-        if (this.fixed) {
-          this._setShadow(this.waterfall && scrollTop > 0);
-        } else {
-          y = Math.min(this._headerDy, Math.max(0, this._oldY + scrollTop - this._oldTop));
-
-          if (this.condenses && this._oldTop >= scrollTop && scrollTop > this._deltaHeight) {
-            y = Math.max(y, this._deltaHeight);
-          }
-          if (this.alwaysVisible) {
-            this._setShadow(this.waterfall && y >= this._deltaHeight);
-          } else {
-            this._setShadow(this.waterfall && y >= this._deltaHeight && y < this._headerDy);
-          }
-
-          if (y !== this._oldY) {
-            this._transformHeader(y);
-            this.fire('app-header-transform', {
-              y: y
-            });
-          }
-        }
+      if (firstSetup || scrollTop === 0) {
+        this.cancelDebouncer('_resize');
+        this._deltaHeight = this._topbar ? this._height - this._topbar.offsetHeight : 0;
+        this._setUpEffect();
+        this._updateContainer(scrollTop, true);
+      } else {
+        this._updateContainer(0, true);
+        this._deltaHeight = this._topbar ? this._height - this._topbar.offsetHeight : 0;
+        this._setUpEffect();
+        // restore the previous scroll state
+        // fix issue in mobile
+        this._updateContainer(this._oldScrollTop, true);
+        this._layoutIfDirty();
+      }
+      // restore no transition
+      this.noTransitions = currentNoTransition;
+    },
+
+    /**
+     *
+     */
+    _updateContainer: function(scrollTop, forceUpdate) {
```

Comment:
I renamed this method to `_updateScrollState`. I believe that name is more consistent with the intend behind it. wdyt?